pkg/causality/internal/node.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/tidwall/btree"
+	"go.uber.org/atomic"
+)
+
+type (
+	workerID = int64
+	nodeID   = int64
+)
+
+const (
+	unassigned = workerID(-1)
+)
+
+var (
+	nextNodeID = atomic.NewInt64(0)
+	nodePool   = &sync.Pool{}
+)
+
+// Node is a node in the dependency graph used
+// in conflict detection.
+type Node struct {
+	id int64 // immutable
+
+	mu             sync.Mutex
+	conflictCounts map[workerID]int
+	// TODO maybe google's btree is better?
+	dependers  btree.Map[nodeID, *Node]
```

Comment:
Why we need a btree.Map here? Or we just need an order map?

Reply:
Yes. An ordered map seems enough. But I have not looked into if there's any production-ready ordered map implementation in Go.

Reply:
Seems one node can only have 0 or 1 dependers?